File: app/src/main/java/com/android/example/popularmovies/ui/detail/DetailViewModelFactory.java
package com.android.example.popularmovies.ui.detail;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.android.example.popularmovies.database.MovieEntry;

public class DetailViewModelFactory extends ViewModelProvider.AndroidViewModelFactory {
    private final MovieEntry mMovieEntry;
    private final Application application;
    private static final Object LOCK = new Object();
    private static DetailViewModelFactory sInstance;

    private DetailViewModelFactory(Application app, MovieEntry movieEntry){
        super(app);
        application = app;
        mMovieEntry = movieEntry;
    }

    public static DetailViewModelFactory getInstance(Application app){
        if (sInstance == null){
            throw new IllegalArgumentException("Instance not created and movie not passed.");
        }
        return sInstance;
    }

    public static DetailViewModelFactory getInstance(Application app, @NonNull MovieEntry movieEntry){
        if (sInstance == null || !sInstance.mMovieEntry.equals(movieEntry)){
            synchronized (LOCK) {
                sInstance = new DetailViewModelFactory(app, movieEntry);
            }
        }
        return sInstance;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(DetailViewModel.class)) {
            return (T) new DetailViewModel(application, mMovieEntry);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/com/android/example/popularmovies/ui/detail/DetailViewModelFactory.kt
package com.android.example.popularmovies.ui.detail

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory
import com.android.example.popularmovies.database.MovieEntry

class DetailViewModelFactory private constructor(private val application: Application, private val mMovieEntry: MovieEntry) : AndroidViewModelFactory(application) {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(DetailViewModel::class.java)) {
            return DetailViewModel(application, mMovieEntry) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }

    companion object {
        private val LOCK = Any()
        private var sInstance: DetailViewModelFactory? = null
        fun getInstance(app: Application?): DetailViewModelFactory? {
            requireNotNull(sInstance) { "Instance not created and movie not passed." }
            return sInstance
        }

        fun getInstance(app: Application, movieEntry: MovieEntry): DetailViewModelFactory? {
            if (sInstance == null || sInstance!!.mMovieEntry != movieEntry) {
                synchronized(LOCK) { sInstance = DetailViewModelFactory(app, movieEntry) }
            }
            return sInstance
        }
    }

}